### PR TITLE
WT-6578 Stop eviction/checkpoint undoing each other's work when running concurrently

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
We've been seeing a similar pattern in a few recent build failures which are centred around concurrent eviction and checkpoint and more specifically, choosing updates for reconciliation that are earlier than the current on-disk value. This causes some updates to get inserted into the history store twice and can erroneously trigger out-of-order timestamp handling and lead to visibility issues.

### CODE REVIEW

**Comment by agorrod** [comment]
@tetsuo-cpp is this ready for review? It failed the latest round of testing, and you haven't asked for a review. So it doesn't look like it is, but I thought I'd ask since it looks like it might be..

**Comment by tetsuo-cpp** [comment]
> @tetsuo-cpp is this ready for review? It failed the latest round of testing, and you haven't asked for a review. So it doesn't look like it is, but I thought I'd ask since it looks like it might be..

I'm preparing it now.

The reason for the failure is because I merged in `develop` and it is not formatted with Clang Format 10 yet.

File: src/reconcile/rec_write.c [hunk 1]
```diff
@@ -529,11 +530,13 @@ __rec_init(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags, WT_SALVAGE_COO
...
      * checkpoints into account.
      */
-    if (WT_IS_METADATA(session->dhandle)) {
+    if (WT_IS_METADATA(session->dhandle))
         WT_ORDERED_READ(ckpt_txn, txn_global->checkpoint_txn_shared.id);
-        if (ckpt_txn != WT_TXN_NONE && WT_TXNID_LT(ckpt_txn, r->last_running))
-            r->last_running = ckpt_txn;
-    }
+    else if (!F_ISSET(r, WT_REC_CHECKPOINT))
+        WT_ORDERED_READ(ckpt_txn, txn_global->checkpoint_txn_shared.pinned_id);
```
Comment (tetsuo-cpp) [inline src/reconcile/rec_write.c:536]: This is the other portion of the change to handle when eviction gets ahead of checkpoint.

In this case, we're going to ensure that if a checkpoint is already running, our `r->last_running` won't go past the checkpoint's pinned timestamp. So our eviction mimick the checkpoint's visibility.

As @quchenhao pointed out, this is only an issue when we have failed evictions. This is normally not an issue because eviction will free updates that got written to the history store or the page so we won't have this problem of checkpoint choosing older updates. However, if we fail an eviction, these updates may be written to the history store and still exist in the chain for the checkpoint to see.

File: test/suite/test_rollback_to_stable10.py [hunk 2]
```diff
@@ -178,7 +178,9 @@ def test_rollback_to_stable(self):
         self.assertGreaterEqual(upd_aborted, 0)
         self.assertGreater(pages_visited, 0)
         self.assertGreaterEqual(hs_removed, 0)
-        self.assertGreater(hs_sweep, 0)
+        # Eviction won't see any of the newer updates that happen concurrently
```
Comment (quchenhao) [inline test/suite/test_rollback_to_stable10.py:181]: Are these comments related to the change?

File: src/history/hs.c [hunk 3]
```diff
@@ -1341,6 +1341,21 @@ __wt_hs_find_upd(WT_SESSION_IMPL *session, WT_ITEM *key, const char *value_forma
                 break;
             }
 
+            /*
```
Comment (tetsuo-cpp) [inline src/history/hs.c:1344]: I'm a bit concerned that we still seem to need this even with the other changes. I'll investigate this more tomorrow.

File: test/suite/test_rollback_to_stable10.py (hunk 2 above)
Comment (tetsuo-cpp) [reply test/suite/test_rollback_to_stable10.py:181]: Yeah, I believe `hs_sweep` corresponds to updates removed from history store as a result of rollback to stable. So since eviction doesn't see those newer updates, there's nothing in the history store that needs to get removed when we perform rollback to stable.

I'll think about how to reword this.

File: src/history/hs.c (hunk 3 above)
Comment (tetsuo-cpp) [reply src/history/hs.c:1344]: Oh, actually this makes sense. When I removed this, I added:

```
+            /* History store records should never be newer than the on-disk value. */
+            WT_ASSERT(session,
+              on_disk_tw->start_ts > hs_start_ts_tmp ||
+                (on_disk_tw->start_ts == hs_start_ts_tmp &&
+                  on_disk_tw->start_txn > hs_cbt->upd_value->tw.start_txn));
```

Anytime a failed eviction with multiple updates newer than the on-disk value, this can happen and it isn't harmful.

File: src/reconcile/rec_visibility.c [hunk 4]
```diff
@@ -370,6 +370,15 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
         return (0);
     }
 
+    /*
+     * Checkpoint may be ahead of eviction. If the newest thing we can see has already been sent to
+     * the history store, we should bail out here.
+     */
+    if (upd != NULL && F_ISSET(upd, WT_UPDATE_HS)) {
```
[the remark this answers is not in the document]
Comment (tetsuo-cpp) [reply src/reconcile/rec_visibility.c:377]: One drawback that @agorrod mentioned it that we can waste a lot of work with this.

File: test/suite/test_rollback_to_stable10.py [hunk 5]
```diff
@@ -306,9 +311,11 @@ def test_rollback_to_stable_prepare(self):
         self.assertEqual(keys_removed, 0)
         self.assertEqual(keys_restored, 0)
         self.assertGreaterEqual(upd_aborted, 0)
-        self.assertGreater(pages_visited, 0)
-        self.assertGreaterEqual(hs_removed, 0)
-        self.assertGreater(hs_sweep, 0)
+        # Since the prepared updates never made their way onto the page, we
```
Comment (kommiharibabu) [inline test/suite/test_rollback_to_stable10.py:314]: The main intention of this test is to check the extra updates that went into the history store but those are not present in the data store as part of the checkpoint when the checkpoint is in-progress. With this change, as it is not possible, do we need this test?

Comment (tetsuo-cpp) [reply test/suite/test_rollback_to_stable10.py:314]: Maybe, do you think we should? I think it's still valuable as a smoke test to race eviction and checkpoint and ensure that the system doesn't trip an assert or die in some other way.

Comment (kommiharibabu) [reply test/suite/test_rollback_to_stable10.py:314]: Yes, we can keep this test to ensure that we don't write anything more in the eviction when checkpoint is in-progress. 
Update the test description with the new intention of the test, so that it will be good.

**Comment by tetsuo-cpp** [comment]
This PR is superseded in favour of #5958.
